projects/grpc-client-server/prober_server/main.go

```diff
@@ -22,26 +22,64 @@ type server struct {
 	pb.UnimplementedProberServer
 }
 
-func (s *server) DoProbes(ctx context.Context, in *pb.ProbeRequest) (*pb.ProbeReply, error) {
-	// TODO: support a number of repetitions and return average latency
-	start := time.Now()
-	_, _ = http.Get(in.GetEndpoint())	// TODO: add error handling here and check the response code
-	elapsed := time.Since(start)
-	elapsedMsecs := float32(elapsed / time.Millisecond)
-
-	return &pb.ProbeReply{LatencyMsecs: elapsedMsecs}, nil
+func (s *server) DoProbes(ctx context.Context, request *pb.ProbeRequest) (*pb.ProbeReply, error) {
+	// get the number of probes from the probe request
+	numberOfProbes := request.GetNumberOfProbes()
+	// fmt.Printf("DoProbes | numberOfProbes %v\n", numberOfProbes)
+
+	// initialise a total time
+	var totalTime time.Duration
+
+	// support a number of repetitions and return average latency
+	for i := 0; i < int(numberOfProbes); i++ {
+		startTime := time.Now()
+		// fmt.Printf("DoProbes | startTime %v\n", startTime)
+
+		// make the request to the endpoint
+		response, err := http.Get(request.GetEndpoint())
+
+		// if the request errors
+		if err != nil {
+			return nil, fmt.Errorf("error: DoProbes request failed: %v", err)
+		}
+
+		// remember to close the response body
+		defer response.Body.Close()
+
+		// if the request status code is not OK
+		if response.StatusCode != http.StatusOK {
+			return nil, fmt.Errorf("error: response was not OK: %v", err)
+		}
+
+		// get the elapsed time
+		elapsedTime := time.Since(startTime)
+		// fmt.Printf("DoProbes | elapsedTime %v\n", elapsedTime)
+
+		// add the elapsed time to the total time
+		totalTime += elapsedTime
+		// fmt.Printf("DoProbes | totalTime %v\n", totalTime)
+	}
+
+	// calculate the average latency in milliseconds
+	// (!) i am deliberately not using milliseconds here, so i can get more precision on the float, but why do i need to do this, this feels janky :/
```

Comment:
You could probably preserve more precision by using one of the methods on `time.Duration` which returns a `float64` and dividing that as appropriate.

There's a question as to how much precision you actually trust, or find useful. Do you know what resolution `time.Now()` offers?

Reply:
No i didn't think about that
I would assume time in Go is in nanoseconds(?)
because Google is such a massive company, and it's their language, and every nanosecond counts for cost/efficiency at scale(?)

from a quick look, `time` uses nanoseconds but it also depends on the hardware(?) unix is 1ms, windows has its own timer.

so is milliseconds enough as a unit for latency to give back to the user?
the reason why i thought about this was for from the lenos of monitoring: a 0.5ms change in latency via prometheus might indicate some change (if it was an internal system?) or is this "too paranoid" of a measurement?

Reply:
I'd generally say milliseconds is probably a pretty reasonable granularity for most users most of the time :) Maybe hundreds of microseconds, _maybe_ even tens of microseconds. Very rarely does anything involving the network happen faster than where tens of microseconds matter - nanoseconds are basically a "within the same machine" kind of granularity rather than an "across the network" matter.